### PR TITLE
feat: support passing index object directly into maybe_set_index

### PR DESCRIPTION
**Pull request by Riik**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue #1239
- Closes #

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below.
[Pandas set_index docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html#pandas-dataframe-set-index)
[Pandas set_index test cases](https://github.com/pandas-dev/pandas/blob/9a015d19514597ad5d1c31f566a0a2bdb17d4cc5/pandas/tests/frame/methods/test_set_index.py#L309)
I've renamed the `column_names: str | list[str]` parameter to `keys: str | Series | list[Series | str]`. I referenced the Pandas docs where which have `keys: label or array-like or list of labels/arrays`. Since this is a backwards-compatability function, I figured it is good for it to be close to the Pandas spec, and the original param name `column` doesn't cover all cases anymore.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
thanks for your PR!

this would break backwards-compatibility, so I don't think we can just rename to `keys` unfortunately

**Comment by Riik** [comment]
Right, so we preserve the `column_names` for backwards comparability, add index, make both optional, and check that they are never both specified. They new signature will then be 
```
def maybe_set_index(df: T, column_names: str | list[str] | None, index: Series | list[Series] | None) -> T
```
I'll give that a try!

File: narwhals/utils.py [hunk 1]
```diff
@@ -297,14 +302,40 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
     df_any = cast(Any, df)
     native_frame = to_native(df_any)
+
+    if column_names is not None and index is not None:
+        msg = "Only one of `column_names` or `keys` should be provided"
+        raise ValueError(msg)
+
+    if not column_names and not index:
+        msg = "Either `column_names` or `keys` should be provided"
+        raise ValueError(msg)
```
Comment (FBruzzesi) [inline narwhals/utils.py:315]:
```suggestion
    if column_names is not None and index is not None:
        msg = "Only one of `column_names` or `index` should be provided"
        raise ValueError(msg)

    if not column_names and not index:
        msg = "Either `column_names` or `index` should be provided"
        raise ValueError(msg)
```

File: narwhals/utils.py [hunk 2]
```diff
@@ -297,14 +302,40 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
+
+        if index is not None:
+            from narwhals.series import Series
+
+            if _is_iterable(index):
+                index = [
+                    key.to_native() if isinstance(key, Series) else key for key in index
+                ]
+            if isinstance(index, Series):
+                index = index.to_native()
```
Comment (FBruzzesi) [inline narwhals/utils.py:333]: How about the following? (will require formatting and maybe ignoring mypy)

```suggestion
        index = [to_native(key, pass_through=True) for key in index] if _is_iterable(index) else [to_native(index, pass_through=True)]
```

**Review by FBruzzesi — COMMENTED** [review]
Hey @Riik thanks for the effort! I left a minor comment and a suggestion to simplify the logic a bit.

Additionally, could we handle the case for which `df` is actually a series? That's the use case I had in mind when I opened the related issue. I.e. if the native_object is a pandas like series, then I would like to modify its index and clearly that's not possible to do unless another Series/Index/iterable is passed.

Since that's the case, I think we should also raise an error if `native_object` is a series but `column_names` is provided

**Comment by Riik** [comment]
> Additionally, could we handle the case for which df is actually a series? That's the use case I had in mind when I opened the related issue. I.e. if the native_object is a pandas like series, then I would like to modify its index and clearly that's not possible to do unless another Series/Index/iterable is passed.
Ah yes, I see the idea now! I think that should work with the current implementation, let me add a test for it

**Comment by Riik** [comment]
@FBruzzesi I've added a test case and also support for setting the index on a Series. It didn't work out of the box since pd.Series ([docs](https://pandas.pydata.org/docs/reference/api/pandas.Series.index.html)) doesn't have a `set_index()` method, but you can set the index directly by assignment. 

Also, there is the extra case where passing in `column_names` in combination with a Series is now invalid. 

I feel like this function is getting a bit unwieldy, i.e. there a lot of conditionals inside the function. Would it be an option to make a separate function `maybe_set_index_series` instead? That would clean up the logic of this function and keep it simpler. Or maybe there is other ways to clean up this function a bit more. Let me know if you have any ideas!

File: narwhals/utils.py [hunk 3]
```diff
@@ -297,14 +302,48 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
+        if isinstance(index, Series):
+            index = index.to_native()
+
+        if is_pandas_like_dataframe(native_frame):
+            return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
+                df_any._compliant_frame._from_native_frame(native_frame.set_index(index))
+            )
+
+        elif is_pandas_like_series(native_frame):
+            native_frame.index = index
```
Comment (FBruzzesi) [inline narwhals/utils.py:342]: @MarcoGorelli pros and cons of setting the index like this vs using [`Series.set_axis`](https://pandas.pydata.org/docs/reference/api/pandas.Series.set_axis.html)?

**Comment by FBruzzesi** [comment]
> @FBruzzesi I've added a test case and also support for setting the index on a Series. It didn't work out of the box since pd.Series ([docs](https://pandas.pydata.org/docs/reference/api/pandas.Series.index.html)) doesn't have a `set_index()` method, but you can set the index directly by assignment.

Thanks! It looks good, I tagged Marco asking if [`Series.set_axis`](https://pandas.pydata.org/docs/reference/api/pandas.Series.set_axis.html) should be prefered instead.
 
> Also, there is the extra case where passing in `column_names` in combination with a Series is now invalid.
> 
> I feel like this function is getting a bit unwieldy, i.e. there a lot of conditionals inside the function. Would it be an option to make a separate function `maybe_set_index_series` instead? That would clean up the logic of this function and keep it simpler. Or maybe there is other ways to clean up this function a bit more. Let me know if you have any ideas!

It still seems pretty reasonable in scope and size. I will take a deeper look tomorrow to see if there is a way to cut some corners :)

**Comment by FBruzzesi** [comment]
@Riik I just merged with main branch and pushed a few changes to:

- parse index and column_names into keys, and avoid having different branches after
- use `set_axis` for series (which I just noticed it makes old version fail - I will get back to `series.index = index`)

**Comment by Riik** [comment]
> @Riik I just merged with main branch and pushed a few changes to parse index and column_names into keys, so that we avoid branching out for dataframe/series in multiple places.

Thank you! I think the function reads quite nicely now, with the `keys` internally it is easier to follow the flow. 

Anything else that still needs to happen on this PR?

**Review by FBruzzesi — APPROVED** [review]
I am (biased and) happy with the state of this! I will leave the last word to Marco for the `set_axis` vs `s.index=index` question before merging!
 
DogHighFiveGIF

File: narwhals/utils.py [hunk 4]
```diff
@@ -274,13 +274,20 @@ def maybe_get_index(obj: T) -> Any | None:
...
+    df: T,
+    column_names: str | list[str] | None = None,
+    *,
+    index: Series | list[Series] | None = None,
+) -> T:
     """
-    Set columns `columns` to be the index of `df`, if `df` is pandas-like.
+    Set the index of `df`, if `df` is pandas-like. The index can either be specified as
+    a existing column name or list of column names with `column_names`, or set directly
+    with a Series or list of Series with `index`.
```
Comment (MarcoGorelli) [inline narwhals/utils.py:286]: the first line of a docstring should generally be a single sentence (separately, we should enable the ruff rule for this, D400)

File: narwhals/utils.py [hunk 5]
```diff
@@ -297,15 +304,41 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
+            df_any._compliant_frame._from_native_frame(native_obj.set_index(keys))
+        )
+    elif is_pandas_like_series(native_obj):
+        if column_names:
+            msg = "Cannot set index using column names on a Series"
+            raise ValueError(msg)
+        native_obj.index = keys
+        return df_any._from_compliant_series(  # type: ignore[no-any-return]
+            df_any._compliant_series._from_native_series(native_obj)
         )
```
Comment (MarcoGorelli) [inline narwhals/utils.py:339]: if pandas doesn't have `Series.set_index`, should we be allowing this at all?

File: narwhals/utils.py [hunk 6]
```diff
@@ -297,15 +304,41 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
...
-            df_any._compliant_frame._from_native_frame(
-                native_frame.set_index(column_names)
-            )
+            df_any._compliant_frame._from_native_frame(native_obj.set_index(keys))
+        )
+    elif is_pandas_like_series(native_obj):
+        if column_names:
+            msg = "Cannot set index using column names on a Series"
+            raise ValueError(msg)
+        native_obj.index = keys
```
Comment (MarcoGorelli) [inline narwhals/utils.py:336]: isn't this going to modify the original object?

File: narwhals/utils.py (hunk 5 above)
Comment (MarcoGorelli) [reply narwhals/utils.py:339]: ah sorry, this is exactly what Francesco had requested, ignore this comment

File: narwhals/utils.py [hunk 7]
```diff
@@ -274,15 +274,37 @@ def maybe_get_index(obj: T) -> Any | None:
...
+    *,
+    index: Series | list[Series] | None = None,
+) -> T:
     """
-    Set columns `columns` to be the index of `df`, if `df` is pandas-like.
+    Set the index of `df`, if `df` is pandas-like, otherwise this is a no-op.
+
+    Arguments:
+        df: object for which maybe set the index (can be either a Narwhals `DataFrame`
+            or `Series`).
```
Comment (MarcoGorelli) [inline narwhals/utils.py:288]: is it a bit odd for this to be called `df` when it can be a dataframe or series?

**Review by MarcoGorelli — COMMENTED** [review]
thanks for updating - implementation looks good, just not sure about the argument name `df`

I did a search on https://github.com/search?q=%22maybe_set_index%28%22+language%3APython&type=code&p=1, and nobody would be affected by such a renaming. I think this is rare enough as a function that we can just change it - the only user I found it scikit-lego, who write
```
nw.maybe_set_index(result, ["fold", "part"])
```
(i.e. they pass the first argument positionally, not by name)

**Review by MarcoGorelli — APPROVED** [review]
thanks @Riik  and @FBruzzesi !

**Comment by rik-vandervlist-sympower** [comment]
🙌 Thanks for the feedback and help, it was a fun PR to work on!

**Pull request merged**